Support legacy webhook endpoints, Update HelloSign, Clubhouse and Librato integrations

Niloth-p opened the pull request:
Fixes:
- Fixes #30779
- Fixes #22731
- [Librato -> AppOptics](https://github.com/zulip/zulip/pull/30788#issuecomment-2218561187)

<details>
<summary>Docs screenshots</summary>

AppOptics
[Doc link](https://zulip.com/integrations/doc/librato)
Before updating the fixture
![image](https://github.com/user-attachments/assets/9933da59-378b-48f1-b4d8-81f05640c6d1)
After the last commit that updates the reported epoch time entries in the fixture
![image](https://github.com/user-attachments/assets/f6a96d26-200d-4fee-8dbd-fc49f9e56c2f)

Dropbox Sign
[Doc link](https://zulip.com/integrations/doc/hellosign)
![image](https://github.com/user-attachments/assets/51e54ca2-a79b-408f-ad85-31768aa53320)

Shortcut
[Doc link](https://zulip.com/integrations/doc/clubhouse)
![image](https://github.com/user-attachments/assets/fe4c3b93-827e-4b0f-b30d-393f36f0ad9d)

</details>

<details>
<summary>Outdated</summary>

**Commit 1**: Retained the previous commit as-is for the review. Will need to squash, and add co-author credits.
**Commit 2**: Updated
	- images: logo, bot avatar, example screenshot
	- name, everywhere except in the **changelog. Should that be changed too?**
	- the documentation to reflect any changes in the setup steps.
No updates were particularly needed for the fixtures (in terms of the fields that we are processing) or the request handling / response logic (Needs a particular response, but is carried over from HelloSign).
**Commit 3**: Just added handling and fixture for the callback test event.

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Niloth-p committed:
webhooks: Add test event handler to Dropbox Sign.

timabbott commented on docs/overview/changelog.md:
```diff
@@ -3859,7 +3859,7 @@ _Released 2017-02-06_
 - Added support for (optionally) using PGRoonga to support full-text
   search in all languages (not just English).
-- Added AppFollow, GitLab, Google Calendar, GoSquared, HelloSign,
+- Added AppFollow, GitLab, Google Calendar, GoSquared, Dropbox Sign,
```
Generally it's not correct to edit 2017 changelog entries due to a product renaming itself.

timabbott commented on zerver/lib/integrations.py:
```diff
@@ -427,7 +427,7 @@ def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
     WebhookIntegration("groove", ["customer-support"], display_name="Groove"),
     WebhookIntegration("harbor", ["deployment", "productivity"], display_name="Harbor"),
-    WebhookIntegration("hellosign", ["productivity", "hr"], display_name="HelloSign"),
+    WebhookIntegration("dropbox sign", ["productivity", "hr"], display_name="Dropbox Sign"),
```
I don't think we usually put spaces in the webhook keys in the first parameter? I think those are intended to always match the filename under `zerver/webhooks/`? 

We might need to do something a bit fancy to have the legacy URL work but call the webhook `dropboxsign`.

timabbott commented on zerver/webhooks/hellosign/doc.md:
```diff
@@ -20,7 +20,7 @@ Configuring the HelloSign integration is easy!
 {!congrats.md!}
 
-![](/static/images/integrations/hellosign/001.png)
+![](/static/images/integrations/Dropbox Sign/001.png)
```
Again, we prefer to avoid spaces in filenames, just use an all-lower-case token (`dropboxsign`) here, matching the webhook path.

timabbott commented:
Oh, the comments above are fixed in your PR; https://zulip.readthedocs.io/en/latest/contributing/continuing-unfinished-work.html is the right way to handle things, which would include squashing your changes into the original author's commit, so we don't have first a buggy/busted original commit and then the fixups. (Or if you basically just started from scratch without actually using their work, just a commit authored by you)

Niloth-p committed:
webhooks: Update HelloSign integration to Dropbox Sign.

Fixes #30779.

Co-authored by: Harsh <hk755063@gmail.com>

Niloth-p committed:
webhooks: Support legacy endpoints for renamed webhooks.

Renamed webhooks can now have more than one URL.
The legacy URLs are computed using the newly added `legacy_names`.
Legacy endpoints are tested for all events too.

Niloth-p commented on zerver/lib/integrations.py:
```diff
@@ -242,8 +248,8 @@ def view(self, request: HttpRequest) -> HttpResponseBase:
     @property
-    def url_object(self) -> URLPattern:
-        return path(self.url, self.view)
+    def url_objects(self) -> list[URLPattern]:
```
Wasn't sure about the best-practice differences between
- creating a new property to return legacy urls separately
- adding a `self.urls`, and modifying `self.url_object` to `self.url_objects`
The latter uses shorter code and both of the fields do not seem to be used anywhere else other than the 2 updated files, so I ran with it.

Niloth-p reviewed (commented):
**Updates:**
- Added a commit (and placed it as the first commit) to support legacy endpoints for renamed webhooks.
- Added legacy_name value for DropboxSign.

**Assumption:** We do not want to support the legacy documentation URL - "/integrations/doc/hellosign".

**Regarding not squashing commits in the previous iteration:**

My intention with separating the commits was purely to facilitate reviewing the work split between authors, since otherwise it wouldn't be possible to do that without reviewing the other PR.
Since I had mentioned that squashing of commits and crediting co-author is yet to be done in both the PR description and the commit message, I thought it was fine.
I can now imagine reviewers reviewing from outside Github, and starting with the code. I'll adapt to that. I'm sorry for any wasted time.

Niloth-p commented on zerver/lib/test_classes.py:
```diff
@@ -2296,6 +2296,7 @@ class WebhookTestCase(ZulipTestCase):
     CHANNEL_NAME: str | None = None
     TEST_USER_EMAIL = "webhook-bot@zulip.com"
     URL_TEMPLATE: str
+    LEGACY_URL_TEMPLATES: list[str] | None = None
```
Many webhooks use URL_TEMPLATE, so created a separate variable for legacy templates.

Niloth-p commented on zerver/lib/test_classes.py:
```diff
@@ -2308,6 +2309,7 @@ def test_user(self) -> UserProfile:
     @override
     def setUp(self) -> None:
         super().setUp()
+        self.legacy_urls: list[str] = []
```
Created this separately from self.url to avoid having to modify the several webhooks that use `self.url` directly.

Niloth-p commented on zerver/lib/test_classes.py:
```diff
@@ -2420,37 +2422,38 @@ def check_webhook(
-            else:
-                raise AssertionError(
-                    "No message was sent. Pass expect_noop=True if this is intentional."
+        for url in [*self.legacy_urls, self.url]:
```
`check_webhook()` now tests all urls supported by a webhook.
Just wrapped inside a for loop.

Niloth-p commented on zerver/lib/test_classes.py:
```diff
@@ -2500,13 +2503,21 @@ def send_and_test_private_message(
 
-    def build_webhook_url(self, *args: str, **kwargs: str) -> str:
-        url = self.URL_TEMPLATE
+    def build_webhook_url(self, *args: str, url_template: str | None = None, **kwargs: str) -> str:
```
`build_webhook_url()` now updates legacy URLs, using recursion.

Since webhooks use it with `self.url = build_webhook_url(...)`, they just update self.url and now they update self.legacy_urls. But, of course, the code does not feel tight enough as the legacy urls are silently updated while the self.url is only returned. But this was the way I could think of to implement this without having to edit every single webhook.
I'm eager to hear better approaches.

Niloth-p commented on zerver/lib/test_classes.py:
```diff
@@ -2500,13 +2503,21 @@ def send_and_test_private_message(
+                self.legacy_urls.append(
+                    self.build_webhook_url(*args, url_template=legacy_url_template, **kwargs)
+                )
+            index = len(self.legacy_urls) // 2
```
In this recursion, for every legacy url template, we build the final url, and append at the end.
So, we end up with 2x elements in the list.
And we prune the first half, once we're done recursing.

I'm not sure this is readable and good code. Improvement suggestions / alternative implementations welcome.

Niloth-p reviewed (commented):
Update:
The legacy URL was previously missing test coverage.
- Modified `test_classes.py` to support legacy urls and run tests for all endpoints.
- Updated Dropbox Sign's `tests.py` to add the 1 extra line needed.

Though we just have 1 webhook with 1 legacy URL in this PR, I verified that the code works for multiple webhooks with multiple legacy URLs.

Niloth-p committed:
webhooks: Update Clubhouse integration to Shortcut.

Fixes #22731.

Niloth-p reviewed (commented):
Added a commit for fixing #22731 Update Clubhouse integration to Shortcut.

Updated Clubhouse integration's images, names (exceptions mentioned in PR comment), doc steps. Retains support for legacy URL.

Appended this commit to this PR as it's dependent on the legacy names feature added in the 1st commit. The number of edited files in this PR though has exploded (Clubhouse/Shortcut has a lot of fixtures), so let me know if it would be convenient for reviewers if I break this into 3 separate PRs - legacy name handling + Dropbox Sign integration + Shortcut integration.

Niloth-p committed:
webhooks: Update Librato integration to AppOptics.

Niloth-p committed:
appoptics webhook: Update date used in fixture used for screenshot.

Replaced the epoch time entries in the fixture, in order to make the
example screenshot appear more recent.

Niloth-p commented:
> it's best not to have 2016 dates there

I've updated that in a separate commit (the last one). We could squash it too, if that fits the narrative.

alya commented:
@laurynmm could you please take a look at this one?

laurynmm reviewed (commented):
@Niloth-p - Thanks for making these updates for these integration name changes!

My main feedback at the moment revolves around the testing changes. I'd be nice to have a URL template that could maybe take a `name` parameter ... or maybe have a separate build urls function for webhooks with legacy names? I'm not really sure what the best answer is.

I think it would be worth a prep refactor commit or two to update existing tests and the `WebhookTestCase` to have a legacy/current webhook name and URL testing vs adding in building the legacy URLs to `build_webhook_url` as is. Maybe best to discuss options on CZO to get feedback from more folks?

Also, we should definitely squash the final commit with the other appoptics commit. Every time we add an image to the repository in GitHub, the size of the repository increases, so we don't want to add an image that's immediately going to be replaced in the next commit if we don't have to.

alya commented:
@Niloth-p Can you fix up the PR description so that PR actually gets marked as fixing the issues it fixed in GitHub?
